oauthproxy.go

```diff
@@ -579,16 +581,43 @@ func (p *OAuthProxy) GetRedirect(req *http.Request) (redirect string, err error)
 		redirect = req.Form.Get("rd")
 	}
 	if !p.IsValidRedirect(redirect) {
-		// Use RequestURI to preserve ?query
-		redirect = req.URL.RequestURI()
-		if strings.HasPrefix(redirect, p.ProxyPrefix) {
-			redirect = "/"
+		// Quirk: On reverse proxies that doesn't have support for
+		// "X-Auth-Request-Redirect" header or dynamic header/query string
+		// manipulation (like Traefik v1 and v2), we can try if the header
+		// X-Forwarded-Host exists or not.
+		if isForwardedRequest(req) && p.ReverseProxy {
```

Comment:
Realized another possible issue with trusting headers. I believe if the client sets `X-Auth-Request-Redirect`, Traefik currently wouldn't remove it. https://github.com/traefik/traefik/blob/3ff83fc1f8281851e56dcc5f18600be8fbb6fe9f/pkg/middlewares/auth/forward.go#L163-L164
I've not tested this yet, but seems this would happen looking over their code.

Only hop-by-hop headers are removed (specified [here](https://github.com/vulcand/oxy/blob/577ef7a2383b14cfef7ea10b29599cb1eb067359/forward/headers.go#L26-L38)).

A header whitelist feature has _just_ been added in the next release candidate.
https://github.com/traefik/traefik/blob/v2.4/docs/content/middlewares/forwardauth.md#authrequestheaders
But for most setups this would be very bleeding edge. So I think we shouldn't assume it being available.

Possibly we need to resort to using the header middleware to always remove it. https://doc.traefik.io/traefik/middlewares/headers/#adding-and-removing-headers
@linuxgemini what do you think?

Reply:
@Beanow See the discussion on https://github.com/oauth2-proxy/oauth2-proxy/pull/957#discussion_r547018534